File: mpp-proiect-java-culbec/MotorcycleContest/Model/src/main/java/Model/Participant.java
package Model;

import java.util.Objects;
import java.util.UUID;

public class Participant extends Person<UUID> {
    private String team = null;
    private Integer engineCapacity = null;

    private Participant() {
        super();
    }

    public String getTeam() {
        return team;
    }

    public Integer getEngineCapacity() {
        return engineCapacity;
    }

    public void setTeam(String team) {
        this.team = team;
    }

    public void setEngineCapacity(Integer engineCapacity) {
        this.engineCapacity = engineCapacity;
    }

    public static class Builder {
        public Builder() {
        }

        private Participant member = new Participant();

        public Builder reset() {
            this.member = new Participant();
            return this;
        }

        public Builder setId(UUID id) {
            this.member.setId(id);
            return this;
        }

        public Builder setFirstName(String firstName) {
            this.member.setFirstName(firstName);
            return this;
        }

        public Builder setLastName(String lastName) {
            this.member.setLastName(lastName);
            return this;
        }

        public Builder setTeam(String team) {
            this.member.team = team;
            return this;
        }

        public Builder setEngineCapacity(Integer engineCapacity) {
            this.member.engineCapacity = engineCapacity;
            return this;
        }

        public Participant build() {
            return member;
        }
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        if (!super.equals(o)) return false;
        Participant that = (Participant) o;
        return Objects.equals(team, that.team) && Objects.equals(engineCapacity, that.engineCapacity);
    }

    @Override
    public int hashCode() {
        return Objects.hash(super.hashCode(), team, engineCapacity);
    }

    @Override
    public String toString() {
        return "Model.Participant{" +
                "team='" + team + '\'' +
                ", engineCapacity=" + engineCapacity +
                "} " + super.toString();
    }
}

File: mpp-proiect-csharp-culbec/Model/Participant.cs
namespace Model;

public class Participant : Person<Guid>
{
    public string? Team { get; set; }
    public int EngineCapacity { get; set; }

    private Participant()
    {

    }

    public new class Builder
    {
        private Participant _member = new();

        public Builder Reset()
        {
            _member = new Participant();
            return this;
        }

        public Builder SetId(Guid id)
        {
            _member.Id = id;
            return this;
        }

        public Builder SetFirstName(string firstName)
        {
            _member.FirstName = firstName;
            return this;
        }

        public Builder SetLastName(string lastName)
        {
            _member.LastName = lastName;
            return this;
        }

        public Builder SetTeam(string team)
        {
            _member.Team = team;
            return this;
        }

        public Builder SetEngineCapacity(int engineCapacity)
        {
            _member.EngineCapacity = engineCapacity;
            return this;
        }

        public Participant Build()
        {
            return _member;
        }
    }

    public new static Builder NewBuilder()
    {
        return new Builder();
    }

    public override bool Equals(object? obj)
    {
        if (ReferenceEquals(null, obj)) return false;
        if (ReferenceEquals(this, obj)) return true;
        return base.Equals(obj) && obj.GetType() == GetType() && Equals(Team, ((Participant)obj).Team) &&
               Equals(EngineCapacity, ((Participant)obj).EngineCapacity);
    }

    public override int GetHashCode()
    {
        return base.GetHashCode() + EqualityComparer<string>.Default.GetHashCode(Team!) +
               EqualityComparer<int>.Default.GetHashCode(EngineCapacity);
    }

    public override string ToString()
    {
        return $"{base.ToString()}, {nameof(Team)}: {Team}, {nameof(EngineCapacity)}: {EngineCapacity}";
    }
}